stone/api.py

```diff
@@ -181,24 +207,27 @@ def add_alias(alias):
         return linearized_aliases
 
     def get_route_io_data_types(self):
+        # type: () -> List[UserDefined]
         """
         Returns a list of all user-defined data types that are referenced as
         either an argument, result, or error of a route. If a List or Nullable
         data type is referenced, then the contained data type is returned
         assuming it's a user-defined type.
         """
-        data_types = set()
+        data_types = set()  # type: Set[UserDefined]
         for route in self.routes:
             for dtype in (route.arg_data_type, route.result_data_type,
                           route.error_data_type):
                 while is_list_type(dtype) or is_nullable_type(dtype):
-                    dtype = dtype.data_type
+                    dtype_as_list = cast(DataTypeList, dtype)
```

Comment:
Sure. 

As a side note, this was done mainly because the base class `DataType` doesn't have a `.data_type` property; perhaps this implies that we should have some typecheckable way to say "this has a concrete .data_type"